### PR TITLE
[project-base] resolved (either fixed or ignored) all the errors reported by phpstan level 4

### PR DESCRIPTION
**Pull request by vitek-rostislav**
| Q             | A
| ------------- | ---
|Description, reason for the PR| In #1344, level of PHPStan was raised to 4 in project base, however, from monorepo there were no errors reported. The problem occured lately, after the project-base was split, see https://travis-ci.org/shopsys/project-base/jobs/581096169
|New feature| No 
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: docs/upgrade/UPGRADE-v8.1.0-dev.md [hunk 1]
```diff
@@ -192,14 +192,14 @@ There you can find links to upgrade notes for other versions too.
...
-- increase your PHPStan level to 4 in your `build.xml` ([#1040](https://github.com/shopsys/shopsys/pull/1040))
+- increase your PHPStan level to 4 in your `build.xml` and address all the reported violations ([#1381](https://github.com/shopsys/shopsys/pull/1381))
     ```diff
   - <property name="phpstan.level" value="1"/>
   + <property name="phpstan.level" value="4"/>
     ```
     - a lot of the possible issues should be already resolved if you followed the previous instruction and ran the `php phing annotations-fix` phing command
     - some of the issues related to class extension need to be addressed manually nevertheless (see the ["Framework extensibility" article](../introduction/framework-extensibility.md#problem-3)) for more information
-    - you need to resolve all the other reported problems (any of them can be ignored in your `phpstan.neon`). You can find inspiration in [#1040](https://github.com/shopsys/shopsys/pull/1040)
+    - you need to resolve all the other reported problems (it is up to you whether you decide to address them directly or add ignores in your `phpstan.neon`). You can find inspiration in [#1381](https://github.com/shopsys/shopsys/pull/1381)
```
Comment (LukasHeinz) [inline docs/upgrade/UPGRADE-v8.1.0-dev.md]: Why did you remove mention of #1040 ? I thought that some inspiration can be also found there, especially for fixing your own features...

File: docs/upgrade/UPGRADE-v8.1.0-dev.md [hunk 2]
```diff
@@ -192,14 +192,14 @@ There you can find links to upgrade notes for other versions too.
 
       <import file="${path.framework}/build.xml"/>
     ```
-- increase your PHPStan level to 4 in your `build.xml` ([#1040](https://github.com/shopsys/shopsys/pull/1040))
+- increase your PHPStan level to 4 in your `build.xml` and address all the reported violations ([#1381](https://github.com/shopsys/shopsys/pull/1381))
```
Comment (LukasHeinz) [inline docs/upgrade/UPGRADE-v8.1.0-dev.md]: I would recommend rephrasing this instruction since there is no obligation or urgency
```suggestion
- for the better quality of code in your project we recommend you to increase your PHPStan level to 4 in your `build.xml` and address all the reported violations ([#1381](https://github.com/shopsys/shopsys/pull/1381))
```

File: docs/upgrade/UPGRADE-v8.1.0-dev.md (hunk 1 above)
Comment (vitek-rostislav) [reply docs/upgrade/UPGRADE-v8.1.0-dev.md]: Because I did not see any added value in linking the old PR. All the current issues with phpstan were addressed in this opened PR (however the truth is I was looking for inspiration #1040)

Comment (LukasHeinz) [reply docs/upgrade/UPGRADE-v8.1.0-dev.md]: My thought is that when implementators want to fix stuff in their features they could use our fixes from #1040 to see what type of errors occurred and how to fix them...

File: project-base/phpstan.neon [hunk 3]
```diff
@@ -2,7 +2,40 @@ parameters:
...
+            message: '#(PHPDoc tag @(param|return) has invalid value (.|\n)+ expected TOKEN_IDENTIFIER at offset \d+)#'
+            path: %currentWorkingDirectory%/tests/ShopBundle/Test/Codeception/_generated/AcceptanceTesterActions.php
+        -
+            message: '#(PHPDoc tag @throws with type .+ is not subtype of Throwable)#'
+            path: %currentWorkingDirectory%/tests/ShopBundle/Test/Codeception/_generated/AcceptanceTesterActions.php
+        # In tests, we often grab services using container->get or access persistent refrences using $this->getReference
+        -
+            message: '#(Property (Shopsys|Tests)\\.+::\$.+ \(.+\) does not accept object.)#'
+            path: %currentWorkingDirectory%/tests/ShopBundle/*
+        # In data fixtures, we often access persistent references using $this->getReference
```
Comment (PetrHeinz) [inline project-base/phpstan.neon]: This look awful lot similar to the rule above (both even mention accessing persistent references). Could they be unified?

File: project-base/phpstan.neon [hunk 4]
```diff
@@ -2,7 +2,40 @@ parameters:
...
+        -
+            message: '#(Property Shopsys\\.+::\$.+ \(Shopsys\\.+\) does not accept object\.)#'
+            path: %currentWorkingDirectory%/src/Shopsys/ShopBundle/DataFixtures/*
+        # A helper methods returning an array of persistent references using $this->getReference
+        -
+            message: '#Method Shopsys\\ShopBundle\\DataFixtures\\ProductDataFixtureReferenceInjector::.+\(\) should return array<.+> but returns array<string, object>\.#'
+            path: %currentWorkingDirectory%/src/Shopsys/ShopBundle/DataFixtures/ProductDataFixtureReferenceInjector.php
+        # In tests, there are helper methods for grabbing services using container->get
+        -
+            message: '#(Method .+::.+\(\) should return .+ but returns (object))#'
```
Comment (PetrHeinz) [inline project-base/phpstan.neon]: This seems a bit broad, as it ignores return values of all methods that should return `object` in test classes. Is that OK?

Also, `symfony/symfony` changed the return value to `object|null` in `v3.4.31` (see https://github.com/shopsys/shopsys/pull/1353). Shouldn't this be reflected in the rule?

File: docs/upgrade/UPGRADE-v8.1.0-dev.md (hunk 1 above)
Comment (PetrHeinz) [reply docs/upgrade/UPGRADE-v8.1.0-dev.md]: Yeah that might help as well...
```suggestion
    - you need to resolve all the other reported problems (it is up to you whether you decide to address them directly or add ignores in your `phpstan.neon`). You can find inspiration in [#1381](https://github.com/shopsys/shopsys/pull/1381) and [#1040](https://github.com/shopsys/shopsys/pull/1040)
```

File: project-base/phpstan.neon (hunk 4 above)
Comment (PetrHeinz) [reply project-base/phpstan.neon]: Maybe it's because the regex matches a subset of the string, i.e. it's not `'#^(Method .+::.+\(\) should return .+ but returns (object))$#'`

File: phpstan.neon [hunk 5]
```diff
@@ -34,37 +34,46 @@ parameters:
...
             path: %currentWorkingDirectory%/project-base/src/Shopsys/ShopBundle/DataFixtures/ProductDataFixtureReferenceInjector.php
+        # In tests, we often grab services using container->get or access persistent refrences using $this->getReference
         -
             message: '#(Property (Shopsys|Tests)\\.+::\$.+ \(.+\) does not accept object\.)#'
             path: %currentWorkingDirectory%/project-base/tests/ShopBundle/*
+        # In tests, we often grab services using container->get or access persistent refrences using $this->getReference
         -
             message: '#(Property (Shopsys|Tests)\\.+::\$.+ \(.+\) does not accept object\|null\.)#'
             path: %currentWorkingDirectory%/project-base/tests/ShopBundle/*
+        # In tests, there are helper methods for grabbing services using container->get
```
Comment (PetrHeinz) [inline phpstan.neon]:
```suggestion
        # In tests, there are helper methods for grabbing services using $container->get() or Codeception modules using $this->getModule()
```

File: phpstan.neon [hunk 6]
```diff
@@ -34,37 +34,46 @@ parameters:
...
         -
-            message: '#(Method .+::.+\(\) should return .+ but returns (object|Codeception\\Module).)#'
+            message: '#(Method .+::.+\(\) should return .+ but returns object.)#'
```
Comment (PetrHeinz) [inline phpstan.neon]: I would rather be more strict:

```suggestion
            message: '#(Method .+::.+\(\) should return .+ but returns (object|Codeception\\Module)\.)#'
```

File: phpstan.neon [hunk 7]
```diff
@@ -34,37 +34,46 @@ parameters:
...
             path: %currentWorkingDirectory%/project-base/tests/ShopBundle/*
+        # Actually, we are setting an array item using "$array[] = $this->getReference"
```
Comment (PetrHeinz) [inline phpstan.neon]:
```suggestion
        # Actually, we are setting an array item using "$array[] = $this->getReference()"
```

File: phpstan.neon [hunk 8]
```diff
@@ -34,37 +34,46 @@ parameters:
...
             message: '#(PHPDoc tag @(param|return) has invalid value (.|\n)+ expected TOKEN_IDENTIFIER at offset \d+)#'
             path: %currentWorkingDirectory%/project-base/tests/ShopBundle/Test/Codeception/_generated/AcceptanceTesterActions.php
         -
             message: '#(PHPDoc tag @throws with type .+ is not subtype of Throwable)#'
             path: %currentWorkingDirectory%/project-base/tests/ShopBundle/Test/Codeception/_generated/AcceptanceTesterActions.php
+        # Actually, the method is called on NodeBuilder which is implementing the interface, however, annotating would worsen the readability of the code in Configuration class
         -
             message: '#Call to an undefined method Symfony\\Component\\Config\\Definition\\Builder\\NodeParentInterface::scalarNode\(\)#'
             path: %currentWorkingDirectory%/project-base/src/Shopsys/ShopBundle/DependencyInjection/Configuration.php
+        # In data fixtures, we often access persistent references using $this->getReference
```
Comment (PetrHeinz) [inline phpstan.neon]:
```suggestion
        # In data fixtures, we often access persistent references using $this->getReference()
```

File: phpstan.neon [hunk 9]
```diff
@@ -34,37 +34,46 @@ parameters:
...
         -
             message: '#(Property Shopsys\\.+::\$.+ \(Shopsys\\.+\) does not accept object\.)#'
             path: %currentWorkingDirectory%/project-base/src/Shopsys/ShopBundle/DataFixtures/*
+        # A helper methods returning an array of persistent references using $this->getReference
```
Comment (PetrHeinz) [inline phpstan.neon]:
```suggestion
        # A helper methods returning an array of persistent references using $this->getReference()
```

File: phpstan.neon [hunk 10]
```diff
@@ -34,37 +34,46 @@ parameters:
...
         -
             message: '#Method Shopsys\\ShopBundle\\DataFixtures\\ProductDataFixtureReferenceInjector::.+\(\) should return array<.+> but returns array<string, object>\.#'
             path: %currentWorkingDirectory%/project-base/src/Shopsys/ShopBundle/DataFixtures/ProductDataFixtureReferenceInjector.php
+        # In tests, we often grab services using container->get or access persistent refrences using $this->getReference
```
Comment (PetrHeinz) [inline phpstan.neon]:
```suggestion
        # In tests, we often grab services using $container->get() or access persistent refrences using $this->getReference()
```

File: phpstan.neon [hunk 11]
```diff
@@ -34,37 +34,46 @@ parameters:
...
         -
             message: '#(Property (Shopsys|Tests)\\.+::\$.+ \(.+\) does not accept object\.)#'
             path: %currentWorkingDirectory%/project-base/tests/ShopBundle/*
+        # In tests, we often grab services using container->get or access persistent refrences using $this->getReference
```
Comment (PetrHeinz) [inline phpstan.neon]:
```suggestion
        # In tests, we often grab services using $container->get() or access persistent references using $this->getReference()
```

File: project-base/phpstan.neon (hunk 3 above)
Comment (vitek-rostislav) [reply project-base/phpstan.neon]: True, fixed in c735d3a7d

File: phpstan.neon [hunk 12]
```diff
@@ -34,37 +34,46 @@ parameters:
         -
             message: '#Access to an undefined property PhpParser\\Node::\$name\.#'
             path: %currentWorkingDirectory%/packages/framework/src/Component/Translation/ConstraintViolationExtractor.php
-        # shopsys/project-base - don't forget to add these rules to phpstan.neon in project-base (only exceptions for level 1)
+        # shopsys/project-base - don't forget to add these rules to phpstan.neon in project-base
+        # Ignore annotations in generated code
         -
             message: '#(PHPDoc tag @(param|return) has invalid value (.|\n)+ expected TOKEN_IDENTIFIER at offset \d+)#'
```
Comment (PetrHeinz) [inline phpstan.neon]: Shouldn't we always end the message pattern with `\.` (if not `$`) for clarity and consistence? Otherwise, the patterns will be too loose.

```suggestion
            message: '#(PHPDoc tag @(param|return) has invalid value (.|\n)+ expected TOKEN_IDENTIFIER at offset \d+\.)#'
```
